### PR TITLE
miri enum discriminant handling: Fix treatment of pointers, better error when it is undef

### PR DESCRIPTION
**Pull request by RalfJung**
r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/interpret/memory.rs [hunk 1]
```diff
@@ -317,10 +321,10 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> Memory<'a, 'mir, 'tcx, M> {
         &self,
         ptr: Pointer<M::PointerTag>,
         size: Size,
-        access: bool
+        check: InboundsCheck,
     ) -> EvalResult<'tcx> {
         // if ptr.offset is in bounds, then so is ptr (because offset checks for overflow)
-        self.check_bounds_ptr(ptr.offset(size, &*self)?, access)
```
Comment (oli-obk) [inline src/librustc_mir/interpret/memory.rs]: This is unrelated to this PR, but I don't think we have a test for the overflow behaviour here.

Comment (RalfJung) [reply src/librustc_mir/interpret/memory.rs]: miri tests some things, not sure if it tests this. Also I do not have a good idea for how to test this in a platform-independent way.

But `ptr.offset` certainly fails on overflow. If rustc had unit tests, we'd just let them cover that...

**Comment by RalfJung** [comment]
@bors r=oli-obk
per comment on Zulip

**Comment by RalfJung** [comment]
Fixed tests.
@bors r=oli-pbk

**Comment by RalfJung** [comment]
Rebased.

@bors r=oli-obk

**Pull request merged**